// ensime-client/release/js/lib/dotensime-utils.js
"use strict";
var lisp = require('./lisp/lisp');
var _ = require('lodash');
var Promise = require('bluebird');
var glob = require('glob');
var swankExtras = require('./lisp/swank-extras');
var sexpToJObject = swankExtras.sexpToJObject;
var file_utils_1 = require('./file-utils');
function readDotEnsime(path) {
    return file_utils_1.readFile(path).then(function (raw) {
        var rows = raw.toString().split(new RegExp("\r?\n"));
        var filtered = rows.filter(function (l) { return l.indexOf(';;') != 0; });
        return filtered.join('\n');
    });
}
function parseDotEnsime(path) {
    // scala version from .ensime config file of project
    return readDotEnsime(path).then(function (dotEnsime) {
        var dotEnsimeLisp = lisp.readFromString(dotEnsime);
        var dotEnsimeJs = sexpToJObject(dotEnsimeLisp);
        var subprojects = dotEnsimeJs[':subprojects'];
        var sourceRoots = _.flattenDeep(_.map(subprojects, function (sp) { return sp[':source-roots']; }));
        var scalaVersion = dotEnsimeJs[':scala-version'];
        var scalaEdition = scalaVersion.substring(0, 4);
        return {
            name: dotEnsimeJs[':name'],
            scalaVersion: scalaVersion,
            scalaEdition: scalaEdition,
            javaHome: dotEnsimeJs[':java-home'],
            javaFlags: dotEnsimeJs[':java-flags'],
            rootDir: dotEnsimeJs[':root-dir'],
            cacheDir: dotEnsimeJs[':cache-dir'],
            compilerJars: dotEnsimeJs[':scala-compiler-jars'],
            dotEnsimePath: path,
            sourceRoots: sourceRoots
        };
    });
}
exports.parseDotEnsime = parseDotEnsime;
// Gives promise of .ensime paths
function allDotEnsimesInPaths(paths) {
    var globTask = Promise.promisify(glob);
    var promises = paths.map(function (dir) {
        return globTask('.ensime', {
            cwd: dir,
            matchBase: true,
            nodir: true,
            realpath: true,
            ignore: '**/{node_modules,.ensime_cache,.git,target,.idea}/**'
        });
    });
    var promise = Promise.all(promises);
    var result = promise.then(function (dotEnsimesUnflattened) {
        var thang = _.flattenDeep(dotEnsimesUnflattened);
        function toObj(path) {
            return { path: path };
        }
        return thang.map(toObj);
    });
    return result;
}
exports.allDotEnsimesInPaths = allDotEnsimesInPaths;
function dotEnsimesFilter(path, stats) {
    !stats.isDirectory() && !_.endsWith(path, '.ensime');
}
exports.dotEnsimesFilter = dotEnsimesFilter;
