// ensime-client/release/js/lib/server-startup/server-startup-utils.js
"use strict";
var path = require('path');
var _ = require('lodash');
var Promise = require('bluebird');
var child_process_1 = require('child_process');
var loglevel = require('loglevel');
var log = loglevel.getLogger('server-startup');
var fs = require('fs');
var file_utils_1 = require('../file-utils');
/**
* Sort monkeys and add tools.jar
*/
function fixClasspath(javaHome, classpathList) {
    var toolsJar = path.join(javaHome, 'lib', 'tools.jar');
    //Sort classpath so any jar containing monkey comes first
    var monkey = new RegExp('monkey');
    var sorter = function (jarPath) { return !monkey.test(jarPath); };
    classpathList.push(toolsJar);
    return _.sortBy(classpathList, sorter).join(path.delimiter);
}
exports.fixClasspath = fixClasspath;
/**
 *  Make an array of java command line args for spawn
 */
function javaArgsOf(classpath, dotEnsime, serverVersion, ensimeServerFlags) {
    if (serverVersion === void 0) { serverVersion = "1.0.0"; }
    if (ensimeServerFlags === void 0) { ensimeServerFlags = ""; }
    var args = ["-classpath", classpath, ("-Densime.config=" + dotEnsime.dotEnsimePath)];
    if (serverVersion <= "2") {
        args.push("-Densime.protocol=jerk");
    }
    if (ensimeServerFlags.length > 0) {
        args.push(ensimeServerFlags); // ## Weird, but extra " " broke everyting
    }
    args.push("org.ensime.server.Server");
    return args;
}
exports.javaArgsOf = javaArgsOf;
function javaCmdOf(dotEnsime) {
    return path.join(dotEnsime.javaHome, 'bin', 'java');
}
exports.javaCmdOf = javaCmdOf;
function spawnServer(javaCmd, args, detached) {
    if (detached === void 0) { detached = false; }
    return child_process_1.spawn(javaCmd, args, { detached: detached });
}
function logServer(pid, cacheDir) {
    fs.exists(cacheDir, function (exists) {
        if (exists) {
            var serverLog = fs.createWriteStream(path.join(cacheDir, "server.log"));
            pid.stdout.pipe(serverLog);
            pid.stderr.pipe(serverLog);
            return pid.stdin.end();
        }
        else {
            fs.mkdir(cacheDir, function (err) {
                return logServer(pid, cacheDir);
            });
        }
    });
}
function startServerFromClasspath(classpath, dotEnsime, serverVersion, serverFlags) {
    if (serverFlags === void 0) { serverFlags = ""; }
    return new Promise(function (resolve, reject) {
        var fixedClasspath = fixClasspath(dotEnsime.javaHome, classpath);
        var cmd = javaCmdOf(dotEnsime);
        var args = javaArgsOf(fixedClasspath, dotEnsime, serverVersion, serverFlags);
        log.debug("Starting Ensime server with " + cmd + " " + _.join(args, " "));
        file_utils_1.ensureExists(dotEnsime.cacheDir).then(function () {
            var pid = spawnServer(cmd, args);
            logServer(pid, dotEnsime.cacheDir);
            resolve(pid);
        });
    });
}
exports.startServerFromClasspath = startServerFromClasspath;
