// ensime-client/release/js/lib/instance/ensime-instance-manager.js
"use strict";
var _ = require('lodash');
/**
 * Takes care of mapping project roots to Ensime clients for multiple Ensime project support under same Atom window
 * This might be supported in vscode too, but currently isn't
 * # TODO: Should use sourdeDirs of .ensime to do mapping of files -> ensime instance
 */
var InstanceManager = (function () {
    function InstanceManager() {
        this.instances = [];
    }
    InstanceManager.prototype.registerInstance = function (instance) {
        this.instances.push(instance);
    };
    InstanceManager.prototype.stopInstance = function (dotEnsime) {
        for (var _i = 0, _a = this.instances; _i < _a.length; _i++) {
            var instance = _a[_i];
            if (instance.rootDir == dotEnsime.rootDir) {
                instance.destroy();
                this.instances = _.without(this.instances, instance);
            }
        }
    };
    // optional running ensime client of scala source path O(n)
    InstanceManager.prototype.instanceOfFile = function (path) {
        return _.find(this.instances, function (instance) {
            return _.startsWith(path, instance.dotEnsime.cacheDir) || instance.isSourceOf(path);
        });
    };
    InstanceManager.prototype.destroyAll = function () {
        _.forEach(this.instances, function (instance) { return instance.destroy(); });
    };
    InstanceManager.prototype.firstInstance = function () {
        return this.instances[0];
    };
    InstanceManager.prototype.isStarted = function (dotEnsimePath) {
        return _.some(this.instances, function (instance) { return instance.dotEnsime.dotEnsimePath == dotEnsimePath; });
    };
    return InstanceManager;
}());
exports.InstanceManager = InstanceManager;
