// @pivotal-tools/atom-languageclient-commons/build/java-process-language-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const portfinder_1 = require("portfinder");
const net_1 = require("net");
const atom_1 = require("atom");
const sts_adapter_1 = require("./sts-adapter");
const atom_languageclient_1 = require("atom-languageclient");
const jvm_launch_utils_1 = require("@pivotal-tools/jvm-launch-utils");
class JavaProcessLanguageClient extends atom_languageclient_1.AutoLanguageClient {
    constructor(serverHome, serverLauncherJar) {
        super();
        this.serverHome = serverHome;
        this.serverLauncherJar = serverLauncherJar;
        this.DEBUG = false;
    }
    getServerJar() {
        return path.resolve(this.serverHome, this.serverLauncherJar);
    }
    showErrorMessage(detail, desc) {
        const atomEnv = atom;
        const notification = atomEnv.notifications.addError('Cannot start Language Server', {
            dismissable: true,
            detail: detail,
            description: desc,
            buttons: [{
                    text: 'OK',
                    onDidClick: () => {
                        notification.dismiss();
                    },
                }]
        });
        return Promise.reject(new Error(detail));
    }
    getInitializeParams(projectPath, process) {
        const initParams = super.getInitializeParams(projectPath, process);
        initParams.capabilities = {
            workspace: {
                executeCommand: {}
            }
        };
        return super.getInitializeParams(projectPath, process);
    }
    startServerProcess(projectPath) {
        // TODO: Remove when debugging is over
        const atomEnv = atom;
        atomEnv.config.set('core.debugLSP', true);
        let childProcess;
        if (this.DEBUG) {
            return this.connectToLS();
        }
        return new Promise((resolve, reject) => {
            let basePort = Math.floor(Math.random() * 10000) + 40000;
            portfinder_1.getPort({ port: basePort }, (err, port) => {
                this.server = net_1.createServer(socket => {
                    this.socket = socket;
                    resolve(childProcess);
                });
                this.server.listen(port, 'localhost', () => {
                    this.launchProcess(port).then(p => childProcess = p);
                });
            });
        });
    }
    connectToLS() {
        return new Promise(resolve => {
            this.socket = net_1.connect({
                port: 5007
            });
            const stdout = {
                setEncoding: () => {
                    return stdout;
                },
                on: () => stdout
            };
            const process = {
                stdin: {},
                stdout: stdout,
                stderr: stdout,
                pid: -1,
                kill: () => {
                    console.log('fake shutdown');
                },
                on: () => process,
                addListener: () => process,
                prependListener: () => process,
                once: () => process,
                prependOnceListener: () => process,
                removeListener: () => process,
                removeAllListeners: () => process,
                setMaxListeners: () => process,
                getMaxListeners: () => 0,
                listeners: () => [],
                emit: () => false,
                eventNames: () => [],
                listenerCount: () => 0
            };
            resolve(process);
        });
    }
    // Start adapters that are not shared between servers
    postInitialization(server) {
        const stsAdapter = this.createStsAdapter() || new sts_adapter_1.StsAdapter();
        server.connection._onRequest({ method: 'sts/moveCursor' }, (params) => stsAdapter.onMoveCursor(params));
        server.connection.onCustom('sts/progress', (params) => stsAdapter.onProgress(params));
        server.connection.onCustom('sts/highlight', (params) => stsAdapter.onHighlight(params));
        server.disposable.add(new atom_1.Disposable(() => {
            if (this.server) {
                this.server.close();
            }
        }));
    }
    preferJdk() {
        return false;
    }
    findJvm() {
        return this.preferJdk() ? jvm_launch_utils_1.findJdk() : jvm_launch_utils_1.findJvm();
    }
    launchProcess(port) {
        return this.findJvm()
            .catch(error => {
            return this.showErrorMessage("Error trying to find JVM", "" + error);
        })
            .then(jvm => {
            if (!jvm) {
                return this.showErrorMessage("Couldn't locate java in $JAVA_HOME or $PATH");
            }
            let version = jvm.getMajorVersion();
            if (version < 8) {
                return this.showErrorMessage('No compatible Java Runtime Environment found', 'The Java Runtime Environment is either below version "1.8" or is missing from the system');
            }
            return this.launchVmArgs(jvm).then(args => {
                args.push(`-Dserver.port=${port}`);
                return this.doLaunchProcess(jvm, this.getServerJar(), args);
            });
        });
    }
    launchVmArgs(jvm) {
        return Promise.resolve([]);
    }
    doLaunchProcess(jvm, launcher, args = []) {
        let vmArgs = args.concat([
            // Atom doesn't have lazy completion proposals support - completionItem/resolve message. Disable lazy completions
            '-Dsts.lsp.client=atom',
            // '-Dlsp.completions.indentation.enable=true', // Looks like Atom has magic indents same like VSCode - comment it out
            '-Dlsp.yaml.completions.errors.disable=true',
        ]);
        this.logger.debug(`starting "${jvm.getJavaExecutable()} ${vmArgs.join('\n')}\n-jar ${launcher}"`);
        return jvm.jarLaunch(launcher, vmArgs, { cwd: this.serverHome });
    }
    createStsAdapter() {
        return null;
    }
}
exports.JavaProcessLanguageClient = JavaProcessLanguageClient;
