// ensime-client/release/js/lib/server-api/server-connection.js
"use strict";
var loglevel = require('loglevel');
var log = loglevel.getLogger('ensime.client');
var Promise = require('bluebird');
var NetworkClient_1 = require('../network/NetworkClient');
function createConnection(httpPort, generalMsgHandler, serverVersion, serverPid) {
    if (serverPid === void 0) { serverPid = undefined; }
    var deferredConnection = Promise.defer();
    var callbackMap = {};
    var ensimeMessageCounter = 1;
    function handleIncoming(msg) {
        var json = JSON.parse(msg);
        log.debug("incoming: ", json);
        var callId = json.callId;
        // If RpcResponse - lookup in map, otherwise use some general function for handling general msgs
        if (callId) {
            try {
                var p = callbackMap[callId];
                log.debug("resolving promise: " + p);
                p.resolve(json.payload);
            }
            catch (error) {
                log.trace("error in callback: " + error);
            }
            finally {
                delete callbackMap[callId];
            }
        }
        else {
            return generalMsgHandler(json.payload);
        }
    }
    function onConnect() {
        deferredConnection.resolve(publicApi());
    }
    function publicApi() {
        log.debug("creating client api");
        return {
            post: post,
            destroy: destroy,
            httpPort: httpPort
        };
    }
    var netClient = new NetworkClient_1.WebsocketClient(httpPort, onConnect, handleIncoming, serverVersion);
    /**
     * Kills server if it was spawned from here.
     */
    function destroy() {
        netClient.destroy();
        if (serverPid)
            serverPid.kill();
    }
    function postString(msg) {
        var p = Promise.defer();
        var wireMsg = "{\"req\": " + msg + ", \"callId\": " + ensimeMessageCounter + "}";
        callbackMap[ensimeMessageCounter++] = p;
        log.debug("outgoing: " + wireMsg);
        netClient.send(wireMsg);
        return p.promise;
    }
    function post(msg) {
        return postString(JSON.stringify(msg));
    }
    return deferredConnection.promise;
}
exports.createConnection = createConnection;
